### PR TITLE
Knepley/fix mesh creation rebased

### PR DESCRIPTION
**Pull request by bishtgautam**

### CODE REVIEW

**Comment by bishtgautam** [comment]
@knepley A bunch of Plex-related MRs were recently merged to PETSc main. Does the current PETSc main preserve the `VecType/MatType` that is set before refinement (e.g. [rdysetup.c#L167-L191](https://github.com/RDycore/RDycore/blob/ccfa405436039cd1bc5b479fe7d93bd055349274/src/rdysetup.c#L167-L191)) after refinement (e.g. [rdysetup.c#L222-L234](https://github.com/RDycore/RDycore/blob/ccfa405436039cd1bc5b479fe7d93bd055349274/src/rdysetup.c#L222-L234))? Or, a new MR is needed to fix the issue for PETSc main?

**Comment by jeff-cohere** [comment]
I'm away from my office currently but can generate a new PETSc Docker image on Monday, provided we know which version to base it on.

**Comment by knepley** [comment]
@bishtgautam The MR that just merged should fix the vectype issue (I hope). I have not tested it yet since I have been in Germany, but if you would and let me know that would be great.

**Comment by bishtgautam** [comment]
@knepley I will test out the latest PETSc main and report back.

**Comment by bishtgautam** [comment]
@knepley, I believe the `VecType/MatType` are now being preserved. Can you look at the logs below and confirm that is the case?

I used the following PETSc's hash:
```
git show -n 1
commit 16ce926e8e1f977214c72b888607cdae2c3b4104 (HEAD -> main, origin/main, origin/HEAD)
Merge: 79e00b32d3a 41c61772ea5
Author: Satish Balay <balay@mcs.anl.gov>
Date:   Fri Nov 10 22:31:50 2023 -0600

    Merge remote-tracking branch 'origin/release'
```

Here are the RDycore logs using this branch for a small problem consisting of 176 cells
- 1 MPI + 1 GPU: [logfile](https://gist.githubusercontent.com/bishtgautam/ff7908ad1aaa1f7119a3de1ddc15a293/raw/3e61e4a06572ac75709b1c009efec76e6e26cfc1/20x10.n_1.pm-gpu-no-debug-gcc-11-2-0-16ce926e8e1.gpu_cuda.18151561.gpu_aware_mpi_1.log)
- 8 MPIs + 8 GPUs: [logfile](https://gist.githubusercontent.com/bishtgautam/5361e1ac02b41f0bed4648e07d54e5a6/raw/9464d4b81ac4915eccd4295d53cfc6d04a427076/20x10.n_8.pm-gpu-no-debug-gcc-11-2-0-16ce926e8e1.gpu_cuda.18151733.gpu_aware_mpi_1.log)

**Comment by knepley** [comment]
Yes, it certainly looks like it to me since 100% of the flops are on the GPU.

**Comment by bishtgautam** [comment]
I'm getting the following warnings on Perlmutter
```
[ 69%] Building C object src/CMakeFiles/rdycore.dir/write_xdmf_output.c.o
/global/cfs/projectdirs/m4267/gbisht/rdycore/src/swe/swe_ceed.c:336:68: warning: argument 3 of type 'PetscReal[3 * boundary.num_edges]' {aka 'double[3 * boundary.num_edges]'} declared with mismatched bound '3 * boundary.num_edges' [-Wvla-parameter]
  336 |                                                          PetscReal boundary_values[3 * boundary.num_edges]) {
      |                                                          ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /global/cfs/projectdirs/m4267/gbisht/rdycore/src/swe/swe_ceed.c:1:
/global/cfs/projectdirs/m4267/gbisht/rdycore/include/private/rdysweimpl.h:34:107: note: previously declared as 'PetscReal[3 * boundary.num_edges]' {aka 'double[3 * boundary.num_edges]'} with bound '3 * boundary.num_edges'
   34 | PETSC_INTERN PetscErrorCode SWEFluxOperatorSetDirichletBoundaryValues(CeedOperator, RDyBoundary boundary, PetscReal[3 * boundary.num_edges]);
      |                                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/global/cfs/projectdirs/m4267/gbisht/rdycore/src/swe/swe_ceed.c:368:78: warning: argument 3 of type 'RDyMaterial[mesh->num_cells]' declared with mismatched bound 'mesh->num_cells' [-Wvla-parameter]
  368 | PetscErrorCode CreateSWESourceOperator(Ceed ceed, RDyMesh *mesh, RDyMaterial materials_by_cell[mesh->num_cells], PetscReal tiny_h,
      |                                                                  ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /global/cfs/projectdirs/m4267/gbisht/rdycore/src/swe/swe_ceed.c:1:
/global/cfs/projectdirs/m4267/gbisht/rdycore/include/private/rdysweimpl.h:36:74: note: previously declared as 'RDyMaterial[mesh->num_cells]' with bound 'mesh->num_cells'
   36 | PETSC_INTERN PetscErrorCode CreateSWESourceOperator(Ceed, RDyMesh *mesh, RDyMaterial[mesh->num_cells], PetscReal, CeedOperator *);
      |                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/global/cfs/projectdirs/m4267/gbisht/rdycore/src/time_series.c:57:82: warning: argument 3 of type 'PetscReal[boundary.num_edges][3]' {aka 'double[boundary.num_edges][3]'} declared with mismatched bound 'boundary.num_edges' [-Wvla-parameter]
   57 | PetscErrorCode AccumulateBoundaryFluxes(RDy rdy, RDyBoundary boundary, PetscReal time_integrated_flux_densities[boundary.num_edges][3]) {
      |                                                                        ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /global/cfs/projectdirs/m4267/gbisht/rdycore/src/time_series.c:1:
/global/cfs/projectdirs/m4267/gbisht/rdycore/include/private/rdycoreimpl.h:196:81: note: previously declared as 'PetscReal[boundary.num_edges][3]' {aka 'double[boundary.num_edges][3]'} with bound 'boundary.num_edges'
  196 | PETSC_INTERN PetscErrorCode AccumulateBoundaryFluxes(RDy, RDyBoundary boundary, PetscReal[boundary.num_edges][3]);
      |                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 71%] Linking C static library librdycore.a
[ 71%] Built target rdycore
```

**Comment by jeff-cohere** [comment]
Does the code run properly? Those warnings looks spurious to me, in the sense that the bounds are not actually mismatched.

**Comment by knepley** [comment]
It is complaining about declaring a variable-length array on the stack. I think this should be malloc'd since the number of boundary edges is probably large.

**Comment by jeff-cohere** [comment]
We are using a VLA array trick here to pass the array size. It's not clear to me that this means the array is allocated on the stack. But I admit VLAs are a tricky area in C.

**Comment by bishtgautam** [comment]
The code runs on Perlmutter without crashing, but I haven't verified if the answer is correct.

**Comment by jeff-cohere** [comment]
If we're not too worried about the warnings for the conformant array parameters on Perlmutter, this PR is ready for review.

File: src/read_config_file.c [hunk 1]
```diff
@@ -614,9 +614,16 @@ static PetscErrorCode ValidateConfig(MPI_Comm comm, RDyConfig *config) {
 static PetscErrorCode SetAdditionalOptions(RDy rdy) {
 #define VALUE_LEN 128
   PetscFunctionBegin;
-  PetscBool has_param;
+  PetscBool has_param, locRefine;
```
Comment (jeff-cohere) [inline src/read_config_file.c:617]: We should use `snake-case` for variables. I can handle this and the other instances of `camelCase`.

File: src/rdysetup.c [hunk 2]
```diff
@@ -574,6 +572,8 @@ static PetscErrorCode InitBoundaries(RDy rdy) {
         const PetscInt *edge_ids;
         PetscCall(ISGetIndices(edge_is, &edge_ids));
         for (PetscInt i = 0; i < num_edges; ++i) {
+          PetscCheck(edge_ids[i] >= e_start, PETSC_COMM_SELF, PETSC_ERR_ARG_WRONG,
```
Comment (jeff-cohere) [inline src/rdysetup.c:575]: We might also add the condition that `edge_ids[i] < e_end`.

**Review by jeff-cohere — APPROVED** [review]
I think this looks good, aside from a couple of very minor issues I can fix. @bishtgautam , do you want anyone else to have a look at this before I make these fixes and merge?

**Comment by bishtgautam** [comment]
@jeff-cohere Please go ahead and make the changes in this PR and merge it. Thanks.

**Comment by jeff-cohere** [comment]
Okay, I've fixed things. When the CI testing finishes without issues, I'll merge this and delete the `kneply/fix-mesh-creation` branch so it doesn't cause confusion.

**Pull request merged**